### PR TITLE
THRIFT-3301: Naming collisions possible in generated Java code

### PR DESCRIPTION
**Pull request by bgould**
Switched to fully-qualified names for all class name references in generated code.  Added test IDL to verify generated code compiles successfully.

### CODE REVIEW

**Comment by jeking3** [comment]
> The following tests FAILED:
> 	 40 - JavaTest (Failed)
> Errors while running CTest

**Comment by nsuke** [comment]
@bgould the failing CI job is make dist tarball verification.
It's typically caused by missing `EXTRA_DIST` entries for newly added files in `Makefile.am`.

**Comment by bgould** [comment]
Thanks @nsuke I'm pretty sure that was the problem.  I created a new IDL file for validating this change, and I've now added it to tests/Makefile.am

**Comment by jeking3** [comment]
Does the cmake install job need to be updated as well?

**Comment by jeking3** [comment]
I'm not seeing why JavaTypes was introduced - was it simply used to manually check the build for output of every native type?  Is the result from JavaTypes compilation used in some unit test that would fail if it didn't produce the correct output?  If not, does it belong in the distribution?

**Comment by bgould** [comment]
@jeking3 

Yes exactly.  I added it to the build process that compiles the java unit tests... so if some change to compiler introduces a regression, the IDL in that file should fail to compile.  I didn't add a unit test for this as I felt that the compilation itself was the test.

Consider the following IDL:

    struct String {
        1: string val
    }

    struct Object {
        1: map<string, String> somemap
    }

It would be possible to make a modification to the compiler that erroneously generates the map type as `java.util.Map<String, String>` instead of `java.util.Map<java.lang.String, String>` or to add some reference in the generated code to `java.lang.Object` that is not fully-qualified... however with the generated classes above in the `thrift.test` package, `javac` should flag the unqualified reference(s) as ambiguous.

**Comment by bgould** [comment]
Also one thing to note is that the Java code generated from following valid IDL will still fail to compile because the names match Java primitive types... I know of no way to avoid this problem:

```thrift
struct int {
  1: i32 val
}

struct boolean {
  1: bool val
}

struct long {
  1: i64 val
}

struct short {
  1: i16 short
}

struct char {
  1: i16 val
}

struct Primitives {
  1: i32 int,
  2: i64 long,
  3: i16 short,
  4: bool boolean,
  5: i16 char
}
```

**Comment by jeking3** [comment]
I think a number of those things are in the "reserved" keyword list for the thrift compiler, so that's okay.  Nice, I'll mark this one approved.

File: lib/java/test/org/apache/thrift/test/TestClient.java [hunk 1]
```diff
@@ -576,7 +581,7 @@ public static void main(String [] args) {
           goodbye.string_thing = "Goodbye4";
           goodbye.byte_thing = (byte)4;
           goodbye.i32_thing = 4;
-          goodbye.i64_thing = (long)4;
+          goodbye.i64_thing = 4;
```
Comment (Jens-G) [inline lib/java/test/org/apache/thrift/test/TestClient.java]: Probably not really an issue, but that change seems unrelated?

File: test/JavaTypes.thrift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+struct Object {
+  1: Integer integer,
+  2: String str,
+  3: Boolean boolean_field,
+  4: Double dbl,
+  5: Byte bite,
+  6: map<i32, Integer> intmap,
+  7: Map somemap,
+}
+
```
Comment (Jens-G) [inline test/JavaTypes.thrift]: Do we need the same for exception?

File: lib/java/test/org/apache/thrift/test/TestClient.java (hunk 1 above)
Comment (bgould) [inline lib/java/test/org/apache/thrift/test/TestClient.java]: Yes that is unrelated... I think my IDE settings are such that it was removed automatically because it was an unncessary... no problem for me to restore it if you prefer.

File: test/JavaTypes.thrift (hunk 2 above)
Comment (bgould) [inline test/JavaTypes.thrift]: that is probably a good idea, will get that added.

Comment (bgould) [inline test/JavaTypes.thrift]: As it happens, adding the Exception type did expose a naming collision... I will track it down and push an update when fixed.

**Comment by bgould** [comment]
@jeking3 @Jens-G 
Looks my last commit caused travis-ci to fail... from what I can tell it doesn't look like it was related to my changes tho (it in the test suite during Haskell generation it looks like)
